File: FactoryMethod/FactoryMethod.cpp
#include<iostream>
using namespace std;

class Animal {
public:
	virtual void attack() = 0;
};

//Concrete Product
class Bird : public Animal {
public:
	void attack() {
		cout << "Chew!" << endl;
	}
};

//Concrete Product
class Cat : public Animal {
public:
	void attack() {
		cout << "Meow!" << endl;
	}
};

//Abstract Factory
class Factory {
public:
	Bird* makeBird() {
		return new Bird();
	}
	Cat* makeCat() {
		return new Cat();
	}
};


//Client call this
void CreatedAndAttack(Factory* factory) {
	Bird* b = factory->makeBird();
	b->attack();

	Cat* c = factory->makeCat();
	c->attack();
}
int main() {
	Factory* f = new Factory();
	CreatedAndAttack(f);
	//Chew！
	//Meow！
}

File: FactoryMethod/FactoryMethod.cs
using System;

interface IAnimal
{
    void attack();
}

class Bird : IAnimal
{
    public void attack()
    {
        Console.WriteLine("Meow!");
    }
}
class Cat : IAnimal
{
    public void attack()
    {
        Console.WriteLine("Chew!");
    }
}


class Factory
{
	public Bird makeBird()
    {
        return new Bird();
    }
    public Cat makeCat()
    {
        return new Cat();
    }
};
class Client
{
    static void CreatedAndAttack(Factory factory)
    {
        Bird b = factory.makeBird();
        b.attack();

        Cat c = factory.makeCat();
        c.attack();
    }
    static void Main()
    {
        Factory f = new Factory();
        CreatedAndAttack(f);
    }
}
